src/workerd/api/streams/standard.c++

```diff
@@ -524,22 +524,26 @@ int getHighWaterMark(const UnderlyingSource& underlyingSource,
 
 // It is possible for the controller state to be released synchronously while
 // we are in the middle of a read. When that happens we need to defer the actual
-// close/error state change until the read call is complete. ReadPendingScope
-// handles this for us. It is constructed when we start a read and destructed
-// synchronously when the call to read completes.
-template <typename C>
-struct ReadPendingScope {
-  jsg::Lock& js;
-  C& controller;
-  ReadPendingScope(jsg::Lock& js, C& controller)
-      : js(js), controller(controller) {
+// close/error state change until the read call is complete. readPendingScope
+// handles this for us.
+template <typename Controller>
+jsg::Promise<ReadResult> readPendingScope(
```

Comment:
Also, can't this be `jsg::Promise<T> ensureAlive(..., <Promise<T>()> callback)`?

Reply:
Perhaps, but I think keeping this as is should be fine.